mythril/laser/ethereum/state.py

```diff
@@ -1,16 +1,82 @@
-from z3 import BitVec, BitVecVal, Solver, ExprRef, sat
+from z3 import BitVec, BitVecVal, BitVecRef, BitVecNumRef, BitVecSort, Solver, ExprRef, Concat, sat, simplify, Array, ForAll, Solver, UGT, Implies
+from z3.z3types import Z3Exception
 from mythril.disassembler.disassembly import Disassembly
 from copy import copy, deepcopy
 from enum import Enum
 from random import randint
+from mythril.laser.ethereum.util import get_concrete_int
 
 from mythril.laser.ethereum.evm_exceptions import StackOverflowException, StackUnderflowException
 
-
 class CalldataType(Enum):
     CONCRETE = 1
     SYMBOLIC = 2
 
+class Calldata:
+
+    def __init__(self, tx_id, starting_calldata: bytes=None):
+        self.tx_id = tx_id
+        if starting_calldata:
+            self._calldata = []
+            self.calldatasize = BitVecVal(len(starting_calldata), 256)
+            self.concrete = True
+        else:
+            self._calldata = Array('{}_calldata'.format(self.tx_id), BitVecSort(256), BitVecSort(8))
+            self.calldatasize = BitVec('{}_calldatasize'.format(self.tx_id), 256)
+            self.concrete = False
+
+        self.starting_calldata = starting_calldata if starting_calldata else []
+
+    def set_global_state(self, state):
+        self.state = state
+        constraints = []
+
+        if self.concrete:
+            for cd_byte in self.starting_calldata:
```

Comment:
What do you think about:
```python
self._calldata = [BitVecVal(calldata_byte, 8) if instanceof(calldata_byte, int) else calldata_byte for calldata_byte in self.starting_calldata]
```

Reply:
Also, why not do this part in the constructor? Seems to me that this is part of the initialization logic

Reply:
I haven't been paying attention to pep8 when coding this... I could swap out my code with a line like that though, but it's much longer than the 79 char limit in pep8.

It's not in the constructor because it is dependent on global state. I like the constraints idea below better.